source/opt/scalar_analysis.h

```diff
@@ -0,0 +1,126 @@
+
+  // Create a unary negative node on |operand|.
+  SENode* CreateNegation(SENode* operand);
+
+  // Creates a subtraction between the two operands by adding |operand_1| to the
+  // negation of |operand_2|.
+  SENode* CreateSubtraction(SENode* operand_1, SENode* operand_2);
+
+  // Create an addition node between two operands.
+  SENode* CreateAddNode(SENode* operand_1, SENode* operand_2);
+
+  // Create a multiply node between two operands.
+  SENode* CreateMultiplyNode(SENode* operand_1, SENode* operand_2);
+
+  // Create a node representing a constant integer.
+  SENode* CreateConstant(int64_t integer);
+
+  // Create a value unknown node, such as a load.
+  SENode* CreateValueUnknownNode(const ir::Instruction* inst);
+
+  // Create a CantComputeNode. Used to exit out of analysis.
+  SENode* CreateCantComputeNode();
+
+  // Construct the DAG by traversing use def chain of |inst|.
+  SENode* AnalyzeInstruction(const ir::Instruction* inst);
+
+  // Simplify the |node| by grouping like terms or if contains a recurrent
+  // expression, rewrite the graph so the whole DAG (from |node| down) is in
+  // terms of that recurrent expression.
+  //
+  // For example.
+  // Induction variable i=0, i++ would produce Rec(0,1) so i+1 could be
+  // transformed into Rec(1,1).
+  //
+  // X+X*2+Y-Y+34-17 would be transformed into 3*X + 17, where X and Y are
+  // ValueUnknown nodes (such as a load instruction).
+  SENode* SimplifyExpression(SENode* node);
+
+  // Add |prospective_node| into the cache and return a raw pointer to it. If
+  // |prospective_node| is already in the cache just return the raw pointer.
+  SENode* GetCachedOrAdd(std::unique_ptr<SENode> prospective_node);
+
+  bool IsLoopInvariant(const ir::Loop* loop, const SENode* node) const;
+
+ private:
+  ir::IRContext* context_;
+
+  // A map of instructions to SENodes. Not every SENode comes from an
+  // instruction however this is used when nodes are created through the analyze
+  // instruction methods.
+  std::map<const ir::Instruction*, SENode*> instruction_map_;
+
+  // Helper functor to allow two unique_ptr to nodes to be compare. Only needed
+  // for the unordered_set implementation.
+  struct NodePointersEquality {
+    bool operator()(const std::unique_ptr<SENode>& lhs,
+                    const std::unique_ptr<SENode>& rhs) const {
+      return *lhs == *rhs;
+    }
+  };
+
+  // Cache of nodes. All pointers to the nodes are references to the memory
+  // managed by they set.
+  std::unordered_set<std::unique_ptr<SENode>, SENodeHash, NodePointersEquality>
+      node_cache_;
+
+  SENode* AnalyzeConstant(const ir::Instruction* inst);
+
+  // Handles both addition and subtraction. If the |is_subtraction| flag is set
+  // then the addition will be op1+(-op2) otherwise op1+op2.
+  SENode* AnalyzeAddOp(const ir::Instruction* add, bool is_subtraction);
+
+  SENode* AnalyzeMultiplyOp(const ir::Instruction* multiply);
+
+  SENode* AnalyzePhiInstruction(const ir::Instruction* phi);
```

Comment:
We prefer the data members to come after the functions: https://google.github.io/styleguide/cppguide.html#Declaration_Order